scipio/src/file_stream.rs

```diff
@@ -909,4 +1269,64 @@ mod test {
         reader.close().await.unwrap();
         expect_specific_error!(reader.close().await, "\"Bad file descriptor (os error 9)\"");
     });
+
+    file_stream_write_test!(write_all, path, _k, filename, file, {
+        let mut writer = StreamWriterBuilder::new(file)
+            .with_buffer_size(128 << 10)
+            .build();
+
+        writer.write_all(&[0, 1, 2, 3, 4]).await.unwrap();
+
+        assert_eq!(writer.current_pos(), 5);
+        writer.close().await.unwrap();
+        assert_eq!(writer.current_pos(), 5);
+
+        let rfile = DmaFile::open(&filename).await.unwrap();
+        assert_eq!(rfile.file_size().await.unwrap(), 5);
+
+        let mut reader = StreamReaderBuilder::new(rfile)
+            .with_buffer_size(1024)
+            .build();
+
+        let mut buf = Vec::new();
+        let x = reader.read_to_end(&mut buf).await.unwrap();
+        assert_eq!(x, 5);
+        check_contents!(buf, 0);
+        reader.close().await.unwrap();
+    });
+
+    file_stream_write_test!(write_multibuffer, path, _k, filename, file, {
+        let mut writer = StreamWriterBuilder::new(file)
+            .with_buffer_size(1024)
+            .with_write_behind(2)
+            .build();
+
+        for i in 0..4096 {
+            assert_eq!(writer.current_pos(), i);
+            writer.write_all(&[i as u8]).await.unwrap();
+        }
+
+        writer.close().await.unwrap();
+
+        let rfile = DmaFile::open(&filename).await.unwrap();
+        assert_eq!(rfile.file_size().await.unwrap(), 4096);
+
+        let mut reader = StreamReaderBuilder::new(rfile)
+            .with_buffer_size(1024)
+            .build();
+
+        let mut buf = [0u8; 4096];
+        reader.read_exact(&mut buf).await.unwrap();
+        check_contents!(buf, 0);
+        reader.close().await.unwrap();
+    });
+
+    file_stream_write_test!(write_close_twice, path, _k, filename, file, {
+        let mut writer = StreamWriterBuilder::new(file)
+            .with_buffer_size(128 << 10)
+            .build();
+
+        writer.close().await.unwrap();
+        expect_specific_error!(writer.close().await, "\"Bad file descriptor (os error 9)\"");
+    });
```

Comment:
It would be nice to have some extra tests. Some ideas:
* assert error propagate well when trying to create a `StreamWriter` with a read-only file
* assert gentle failure when creating an invalid `StreamWriter` (`with_write_behind(0)` and/or `with_buffer_size(0)`)
* assert we fail when issuing multiple writes from different `StreamWriter`

Reply:
Those comments probably apply to the `StreamReader` to an extent as well.

Reply:
The first two suggestions are good.
The Third, I fail to see the point. It is indeed possible, although weird, to use the same file for two `StreamWriter`. How could we possibly defend against that?

For readers it is more than that: you actually may want many stream readers into the same file.

Reply:
About the second suggestion, though, I think those things should be considered hints and not hard values. If you pass zero the implementation should round them up to a reasonable number. Same thing with buffer sizes that are not powers of two.

However that still warrants a test.

Reply:
For 3, I was referring to this:
https://github.com/DataDog/scipio/pull/68/commits/eb26095df5df79e4ccb9253648bc41d3c60b7a8a#diff-b7113f50e181897a934ae801f9253b34R719-R723

How do we trigger this in a unit test?

Reply:
I don't think we can. That's an assert condition, it should not be possible unless we coded something wrong. Seems beyond the scope of a unit test since if I knew I had coded anything wrong, I would have fixed it ! =)